extensions/amp-carousel/0.1/build-dom.js

```diff
@@ -0,0 +1,356 @@
+    if (!elements[i]) {
+      throw new Error('Invalid server render');
+    }
+  }
+}
+
+/**
+ * Builds a carousel button for next/prev.
+ * @param {!Element} element
+ * @param {{className: string, title: string, enabled: boolean}} options
+ * @return {?HTMLDivElement}
+ */
+function buildButton(element, {className, enabled, title}) {
+  /**
+   * In scrollable carousel, the next/previous buttons add no functionality
+   * for screen readers as scrollable carousel is just a horizontally
+   * scrollable div which ATs navigate just like any other content.
+   * To avoid confusion, we therefore set the role to presentation for the
+   * controls in this case.
+   */
+  const ariaRole = getType(element) === 'scroll' ? 'presentation' : 'button';
+
+  const button = element.ownerDocument.createElement('div');
+  button.setAttribute('tabindex', '0');
+  button.classList.add(ClassNames.BUTTON, className);
+  button.setAttribute('role', ariaRole);
+  button.setAttribute('title', title);
+  setButtonState(button, enabled);
+  element.appendChild(button);
+  return button;
+}
+
+/**
+ *
+ * @param {!HTMLDivElement} button
+ * @param {boolean} enabled
+ */
+export function setButtonState(button, enabled) {
+  button.classList.toggle('amp-disabled', !enabled);
+  button.setAttribute('aria-disabled', String(!enabled));
+  button.setAttribute('tabindex', String(enabled ? 0 : -1));
+}
+
+/**
+ * Builds the DOM necessary for amp-carousel.
+ * @param {!Element} element
+ * @param {number} slideCount
+ * @return {{
+ *   prevButton: !HTMLDivElement,
+ *   nextButton: !HTMLDivElement
+ * }}
+ */
+export function buildCarouselControls(element, slideCount) {
+  if (isServerRendered(element)) {
+    return queryCarouselControls(element);
+  }
+
+  const doc = element.ownerDocument;
+  if (isAmp4Email(doc) || element.hasAttribute('controls')) {
+    element.classList.add(ClassNames.HAS_CONTROL);
+  }
+
+  const hasLoop = element.hasAttribute('loop');
+  const prevIndex = hasLoop ? slideCount : 0;
+  const nextIndex = slideCount > 1 ? 2 : hasLoop ? 0 : 1;
+  const prevButton = buildButton(element, {
+    className: ClassNames.PREV_BUTTON,
+    title: getPrevButtonTitle(element, {
+      index: String(prevIndex),
+      total: String(slideCount),
+    }),
+    enabled: element.hasAttribute('loop'),
+  });
+  const nextButton = buildButton(element, {
+    className: ClassNames.NEXT_BUTTON,
+    title: getNextButtonTitle(element, {
+      index: String(nextIndex),
+      total: String(slideCount),
+    }),
+    enabled: slideCount > 1,
+  });
+  return {prevButton, nextButton};
+}
+
+/**
+ * Queries for all of the necessary DOM Elements to assign to ivars
+ * @param {!Element} element
+ * @return {{
+ *   prevButton: !HTMLDivElement,
+ *   nextButton: !HTMLDivElement
+ * }}
+ */
+export function queryCarouselControls(element) {
+  const prevButton = /** @type {!HTMLDivElement} */ (
+    element.querySelector(`.${escapeCssSelectorIdent(ClassNames.PREV_BUTTON)}`)
```

Comment:
This is becoming a very common pattern in these queryXXX functions. Might be useful to add something to `#core/dom/query` a la 
```
// whatever the naming convention is for those functions
function queryByClassName(element, className) {
  return element.querySelector(`.${escapeCssSelectorIdent(className)}`);
}
```

Reply:
Mind if I bump this idea to a followup PR which implements `queryByClass` and also promotes `assertDomQueryResults`?